packages/ens-utils/src/registrar.ts

```diff
@@ -0,0 +1,217 @@
+
+  /** Extend an existing registration. */
+  Renew: "renew",
+} as const;
+
+export type RegistrarAction = (typeof RegistrarAction)[keyof typeof RegistrarAction];
+
+/**
+ * The type of registrar charge that is being applied to the domain.
+ */
+export const RegistrarChargeType = {
+  /** The base fee for the name. */
+  BaseFee: "base-fee",
+
+  /** A fee charged for names with special attributes. */
+  SpecialNameFee: "special-name-fee",
+
+  /** A temporary fee charged after a name has been recently released. */
+  TemporaryFee: "recently-released-fee",
+
+  /** A service fee. */
+  ServiceFee: "service-fee",
+} as const;
+
+export type RegistrarChargeType = (typeof RegistrarChargeType)[keyof typeof RegistrarChargeType];
+
+/**
+ * A single distinct charge that must be paid to a `Registrar` to register or
+ * renew a domain.
+ */
+export interface AbstractRegistrarCharge {
+
+  /**
+   * The type of `RegistrarCharge` that is being applied by the `Registrar` to
+   * rent the domain.
+   */
+  type: RegistrarChargeType;
+
+  /**
+   * The price of the `RegistrarCharge` that must be paid to the `Registrar` to
+   * rent the domain.
+   */
+  price: Price;
+
+  /**
+   * The reason why the `RegistrarCharge` is being applied by the `Registrar`
+   * to rent the domain.
+   */
+  reason?: string;
+
+  /**
+   * The period of time that the `RegistrarCharge` is valid for.
+   * 
+   * If `null`, the `RegistrarCharge` is not guaranteed to be time-limited.
+   */
+  validity?: TimePeriod;
+}
+
+export interface RegistrarBaseFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.BaseFee;
+  price: Price;
+  reason?: undefined;
+  validity?: undefined;
+}
+
+export interface RegistrarSpecialNameFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.SpecialNameFee;
+  price: Price;
+  reason: string;
+  validity?: undefined;
+}
+
+export interface RegistrarTemporaryFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.TemporaryFee;
+  price: Price;
+  reason: string;
+  validity: TimePeriod;
+}
+
+export interface RegistrarServiceFee extends AbstractRegistrarCharge {
+  type: typeof RegistrarChargeType.ServiceFee;
+  price: Price;
+  reason?: string;
+  validity?: undefined;
+}
+
+export type RegistrarCharge = RegistrarBaseFee | RegistrarSpecialNameFee | RegistrarTemporaryFee | RegistrarServiceFee;
+
+/**
+ * A price quote from a `Registrar` to register or renew a domain for a given
+ * period of time.
+ */
+export interface AbstractRegistrarPriceQuote {
+
+  /**
+   * The action associated with this `RegistrarPriceQuote`.
+   */
+  action: RegistrarAction;
+
+  /**
+   * The `TimePeriod` that this `RegistrarPriceQuote` is for.
+   */
+  rentalPeriod: TimePeriod;
+
+  /**
+   * The set of distinct `RegistrarCharge` values that must be paid to the
+   * `Registrar` to rent the domain for the period of `rentalPeriod`.
+   * 
+   * May be empty if the domain is free to rent for `rentalPeriod`.
```

Comment:
Is this case possible?

Reply:
In reality

Reply:
Sure. For example: cb.id or uni.eth subnames.

We want to be building towards support for these now and not needing to fundamentally change everything in our code once we expand support officially.

Reply:
Nicely thought!